contracts/SpokePool.sol

```diff
@@ -147,19 +179,88 @@ abstract contract SpokePool is Testable, Lockable, MultiCaller {
+        address sender,
+        address recipient
+    ) public {
+        // We limit the relay fees to prevent the user spending all their funds on fees.
+        require(relayerFeePct <= 0.5e18 && realizedLpFeePct <= 0.5e18, "invalid fees");
+
+        // Use relay count as unique ID for new relay.
+        require(relays[numberOfRelays].relayAmount == 0, "Relay exists");
+        relays[numberOfRelays] =
+            RelayData(
+                recipient,
+                relayToken,
+                relayerFeePct,
+                realizedLpFeePct,
+                amount, // total relay amount
+                0 // total amount filled
+            );
+            
+        emit InitiatedRelay(
+            originChainId, 
+            amount,
+            depositId, 
+            numberOfRelays,  
+            relayerFeePct, 
+            realizedLpFeePct, 
+            relayToken, 
+            sender, 
+            recipient
+        );
+
+        // Questions:
+        // - do we need to store relayer (i.e. msg.sender) anywhere on-chain? Or can we assume data worker
+        //   can grab it from event data.
+
+        numberOfRelays += 1;
+    }
 
     function fillRelay(
-        uint256 relayId,
-        uint256 fillAmount,
+        uint64 relayId,
+        uint256 amount,
         uint256 repaymentChain
-    ) public {}
+    ) public {
+        RelayData memory relay = relays[relayId];
+        // The following check will fail if:
+        // - relay has not been instantiated and relayAmount = 0.
+        // - caller's desired amount to fill would send filledAmount over relayAmount.
+        require(
+            amount > 0 &&
+            relay.relayAmount >= relay.filledAmount + amount, 
+            "Invalid remaining relay amount"
+        );
+
+        // Update total filled amount with this new fill. Each fill can be uniquely identified by the
+        // total filled amount including itself, the relay ID, and the chain ID of this contract.
+        relays[relayId].filledAmount += amount;
+
+        // Pull fill amount net fees from caller, which is the amount owed to the recipient.  The relayer will receive 
+        // this amount plus the relayer fee after the relayer refund is processed.
+        uint256 amountNetFees = amount - _getAmountFromPct(
+            relay.realizedLpFeePct + relay.relayerFeePct,
+            amount
+        );
+        IERC20(relay.relayToken).safeTransferFrom(msg.sender, relay.recipient, amountNetFees);
+
+        // If relay token is weth then unwrap and send eth.
+        if (relay.relayToken == wethAddress) {
+            _unwrapWETHTo(payable(relay.recipient), amountNetFees);
+            // Else, this is a normal ERC20 token. Send to recipient.
+        } else IERC20(relay.relayToken).safeTransfer(relay.recipient, amountNetFees);
+
+        // Questions:
+        // - same question as above: do we need to store relayer (i.e. msg.sender) anywhere on-chain? Or can we assume data worker
```

Comment:
I think including it in the event is best. something like `filler` or `relayFiller`